providers/base/bin/prime_offload_tester.py

```diff
@@ -0,0 +1,334 @@
+    NO_CARD_ID = -1
+    NO_CARD_NAME = -2
+    OFFLOAD_FAIL = -3
+    NOT_SUPPORT_NV_PRIME = -4
+    NV_DRIVER_ERROR = -5
+
+
+class PrimeOffloader:
+    """
+    A class used to execute process to specific GPU.
+    Have to run this as root.
+
+    :attr logger: console logger
+    :type logger: obj
+
+    :attr check_result:
+        store the result of checking offloading is ok or not.
+    :type check_result: int
+    """
+
+    logger = logging.getLogger()
+
+    check_result = PrimeOffloaderError.OFFLOAD_FAIL
+
+    def find_card_id(self, pci_name):
+        """
+        use pci name to find card id under /sys/kernel/debug/dri
+
+        :param pci_name: pci device name in NNNN:NN:NN.N format
+        :type card_name: str
+        """
+        cmd = "sudo grep -lr --include=name \"{}\""  \
+              " /sys/kernel/debug/dri 2>/dev/null".format(pci_name)
+        try:
+            if not re.match("[0-9]{4}:[0-9]{2}:[0-9]{2}.[0-9]", pci_name):
+                return PrimeOffloaderError.NO_CARD_ID
+
+            card_path = subprocess.run(cmd, shell=True,
+                                       stdout=subprocess.PIPE,
+                                       universal_newlines=True)
+            card_id = card_path.stdout.split('/')[5]
+            return card_id
+        except (ValueError, IndexError, AttributeError):
+            return PrimeOffloaderError.NO_CARD_ID
+
+    def find_card_name(self, pci_name):
+        """
+        use pci name to find card name by lshw
+
+        :param pci_name: pci device name in NNNN:NN:NN.N format
+        :type card_name: str
+        """
+        cmd = 'sudo lshw -c display -json'
+        try:
+            card_infos = subprocess.run(cmd, shell=True,
+                                        stdout=subprocess.PIPE,
+                                        universal_newlines=True)
+            infos = json.loads(card_infos.stdout)
+            for info in infos:
+                if pci_name in info['businfo']:
+                    return info['product']
+        except (ValueError, IndexError, TypeError):
+            return None
+
+    def check_offload(self, cmd, card_id, card_name):
+        """
+        Use to check provided command is executed on specific GPU.
+
+        .. note::
+            While setting prime offload environment such as DRI_PRIME,
+            the process will be listed under kernel debug interface.
+            The location of kernel debug interface is
+            /sys/kernel/debug/dri/<card id>,
+            and the process could be found in
+            /sys/kernel/debug/dri/<card id>/clients
+
+        :param cmd: command that running under prime offload
+        :type cmd: str
+
+        :param card_id: card id of dri device
+        :type card_id: str
+
+        :param card_name: card name of dri device
+        :type card_name: str
+        """
+        cmd_without_args = cmd.split(' ')[0]
+        for index in range(1, 11):
+            time.sleep(2)
```

Comment:
What is this sleep for?
Why two seconds?
If we need to wait for some condition to occur, let's check for that in a busy loop with a tiny sleep (like 10ms) instead of arbitrary waiting for some time.

Reply:
This sleep could make sure it won't use too much CPU resources. If we use a busy loop and the timeout is very long, it might make some issues.